osutil/sync_dir.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+}
+
+var errSameState = fmt.Errorf("file state has not changed")
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions are what is desired. Unexpected files are
+// removed. Missing files are created and differing files are corrected.  Files
+// not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
```

Comment:
That's not what I meant, and it sounds very dangerous to do that.

What I meant is all or nothing. On any errors, we bail out and remove all known content. This is the only sane thing to do here, IMO, because nobody has ever pondered about the consequences of partially applying an arbitrary subset of security profiles. It would be basically closing our eyes and hoping for the best, which is not a sane strategy with security.

By backing out and removing all known content we ensure that either what we asked for is accomplished, or nothing else is.